client/src/main/java/io/pravega/client/stream/impl/ControllerImpl.java

```diff
@@ -131,20 +145,51 @@ public ControllerImpl(final URI controllerURI, final ControllerImplConfig config
      * @param channelBuilder The channel builder to connect to the service instance.
      * @param config         The configuration for this client implementation.
      * @param executor       The executor service to be used internally.
+     * @param creds          The credentials if any.
+     * @param enableTls      The flag to turn on TLS for controller interactions.
+     * @param tlsCertFile    File containing the public certificate for TLS.
      */
     @VisibleForTesting
     public ControllerImpl(ManagedChannelBuilder<?> channelBuilder, final ControllerImplConfig config,
-                          final ScheduledExecutorService executor) {
+                          final ScheduledExecutorService executor, PravegaCredentials creds,
+                          boolean enableTls, String tlsCertFile) {
         Preconditions.checkNotNull(channelBuilder, "channelBuilder");
         this.executor = executor;
+        this.creds = creds;
         this.retryConfig = Retry.withExpBackoff(config.getInitialBackoffMillis(), config.getBackoffMultiple(),
                 config.getRetryAttempts(), config.getMaxBackoffMillis())
-                .retryingOn(StatusRuntimeException.class)
-                .throwingOn(Exception.class);
-
+                                .retryingOn(StatusRuntimeException.class)
+                                .throwingOn(Exception.class);
+
+        if (enableTls) {
+            SslContextBuilder sslContextBuilder = null;
+            if (!Strings.isNullOrEmpty(tlsCertFile)) {
+                sslContextBuilder = GrpcSslContexts.forClient().trustManager(new File(tlsCertFile));
+            } else {
+                sslContextBuilder = GrpcSslContexts.forClient();
+            }
+            try {
+                channelBuilder = ((NettyChannelBuilder) channelBuilder).sslContext(sslContextBuilder.build())
+                                                                       .negotiationType(NegotiationType.TLS);
+            } catch (SSLException e) {
+                throw new CompletionException(e);
+            }
+        } else {
+            channelBuilder = ((NettyChannelBuilder) channelBuilder).negotiationType(NegotiationType.PLAINTEXT);
+        }
         // Create Async RPC client.
         this.channel = channelBuilder.build();
         this.client = ControllerServiceGrpc.newStub(this.channel);
+        if (creds != null) {
+            PravegaCredsWrapper wrapper = new PravegaCredsWrapper(creds);
+            this.client = client.withCallCredentials(MoreCallCredentials.from(wrapper));
```

Comment:
ok so this is why you have made client non-final.. 
can we use builder for client? so that in the end we simply do clientBuilder.build
that way we can keep the client "final".

Reply:
I have couple more comments saying similar things. Will log an issue on the client side for this. I do not think it should be part of this PR.